osutil/io.go

```diff
@@ -62,53 +107,123 @@ func AtomicWriteFileChown(filename string, data []byte, perm os.FileMode, flags
 	}
 	tmp := filename + "." + strutil.MakeRandomString(12)
 
-	// XXX: if go switches to use aio_fsync, we need to open the dir for writing
-	dir, err := os.Open(filepath.Dir(filename))
+	fd, err := os.OpenFile(tmp, os.O_WRONLY|os.O_CREATE|os.O_TRUNC|os.O_EXCL, perm)
 	if err != nil {
-		return err
+		return nil, err
 	}
-	defer dir.Close()
 
-	fd, err := os.OpenFile(tmp, os.O_WRONLY|os.O_CREATE|os.O_TRUNC|os.O_EXCL, perm)
-	if err != nil {
-		return err
+	return &atomicFile{
+		File:    fd,
+		target:  filename,
+		tmpname: tmp,
+		uid:     uid,
+		gid:     gid,
+	}, nil
+}
+
+// ErrCannotCancel means the Commit operation failed at the last step, and
+// your luck has run out.
+var ErrCannotCancel = errors.New("cannot cancel: file has already been renamed")
+
+func (aw *atomicFile) Close() error {
+	aw.closed = true
+	return aw.File.Close()
+}
+
+func (aw *atomicFile) Cancel() error {
+	if aw.renamed {
+		return ErrCannotCancel
 	}
-	defer func() {
-		e := fd.Close()
-		if err == nil {
-			err = e
-		}
-		if err != nil {
-			os.Remove(tmp)
-		}
-	}()
 
-	// according to the docs, Write returns a non-nil error when n !=
-	// len(b), so don't worry about short writes.
-	if _, err := fd.Write(data); err != nil {
-		return err
+	var e1, e2 error
+	if aw.tmpname != "" {
+		e1 = os.Remove(aw.tmpname)
 	}
+	if !aw.closed {
+		e2 = aw.Close()
+	}
+	if e1 != nil {
+		return e1
+	}
+	return e2
+}
 
-	if uid > -1 && gid > -1 {
-		if err := fd.Chown(uid, gid); err != nil {
+var chown = (*os.File).Chown
+
+func (aw *atomicFile) Commit() error {
+	if aw.uid > -1 && aw.gid > -1 {
+		if err := chown(aw.File, aw.uid, aw.gid); err != nil {
 			return err
 		}
-	} else if uid > -1 || gid > -1 {
-		return errors.New("internal error: AtomicWriteFileChown needs none or both of uid and gid set")
 	}
 
+	var dir *os.File
 	if !snapdUnsafeIO {
-		if err := fd.Sync(); err != nil {
+		// XXX: if go switches to use aio_fsync, we need to open the dir for writing
+		d, err := os.Open(filepath.Dir(aw.target))
+		if err != nil {
+			return err
+		}
+		dir = d
+		defer dir.Close()
+
+		if err := aw.Sync(); err != nil {
 			return err
 		}
 	}
 
-	if err := os.Rename(tmp, filename); err != nil {
+	if err := aw.Close(); err != nil {
+		return err
+	}
+
+	if err := os.Rename(aw.tmpname, aw.target); err != nil {
 		return err
 	}
+	aw.renamed = true // it is now too late to Cancel()
 
 	if !snapdUnsafeIO {
 		return dir.Sync()
```

Comment:
as I said one option would be to also remember this error and return on it Cancel as well if not nil

Reply:
Agreed, but then the caller would need to be able to distinguish "got weird os error from Cancel because Commit failed" from "got weird os error from Cancel because Cancel failed", which would involve a whole lot of wrapping errors; doesn't seem worth it. This way you get a weird os error from commit, and a CannotCancel error from Cancel; you know you're screwed.